File: 2017/5.py
import math 
import functools
import re
import random
import collections

if __name__ == "__main__":
    with open('./input/5.txt') as f:
        data = f.read().splitlines()
        data = list(map(int, data))

        pos = 0
        steps = 0
        while pos >= 0 and pos < len(data):
            dist = data[pos]
            oldpos = pos
            pos += dist
            if data[oldpos] >= 3:
                data[oldpos] -= 1
            else:
                data[oldpos] += 1
            steps += 1
        
        print(steps)
        

File: 2018/5/5.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strings"
	"unicode"
)

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func readFile() string {
	f, err := os.Open("input-5.txt")
	if err != nil {
		fmt.Println(err)
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)

	var data string

	for scanner.Scan() {
		data = scanner.Text()
	}
	
	return data
}

func collapse(input string) string {
	for {
		oldLen := len(input)
		for c := 'a'; c <= 'z'; c++ {
			input = strings.Replace(input, string([]rune{c, unicode.ToUpper(c)}), "", -1) 
			input = strings.Replace(input, string([]rune{unicode.ToUpper(c), c}), "", -1) 

		}
		if oldLen == len(input) {
			break
		}
	}
	return input
}

func main() {
	// read the input string
	data := readFile()
	input := data

	// part 1
	fmt.Printf("Part 1: %d\n", len(collapse(input)))

	// part 2
	smallest := math.MaxInt32
	for c := 'a'; c <= 'z'; c++ {
		newString := strings.Clone(input)
		newString = strings.Replace(newString, string(c), "", -1)
		newString = strings.Replace(newString, string(unicode.ToUpper(c)), "", -1)
		smallest = min(smallest, len(collapse(newString)))
	}
	fmt.Printf("Part 2: %d\n", smallest)
}